### PR TITLE
Removes Mutexing of individual pins.

### PR DESCRIPTION
**Pull request by tcr**
This PR gets us closer to the APIs in #29. It removes the mutexing of individual pins and explores a Port -> (I2C, Gpio) and Gpio -> Pins space. The Accelerometer and (new) Climate module code consume the entire port accordingly.

This PR also includes a simple web demo to show mutexing a sensor in action.

### CODE REVIEW

File: climate-si7020/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    loop {
+        // Toggle each LED
+        tessel.led[2].toggle().unwrap();
+        tessel.led[3].toggle().unwrap();
+        // Re-execute the loop after sleeping for 100ms
+        sleep(Duration::from_millis(100));
+
+        println!("temperature: {:?}", climate.read_temperature());
+
+        let _ = std::io::stdout().flush();
```
Comment (Dr-Emann) [inline climate-si7020/src/main.rs]: stdout should be line buffered already. I'm not sure this line will actually do anything.

Comment (tcr) [reply climate-si7020/src/main.rs]: Debugging code >.< Thanks for catching this @Dr-Emann

File: tessel/src/lib.rs [hunk 2]
```diff
@@ -82,86 +78,146 @@ pub struct PortGroup {
...
-pub struct Pin<'a> {
-    index: usize,
-    _guard: MutexGuard<'a, ()>,
-    socket: Arc<Mutex<PortSocket>>,
-}
+impl Port {
+    pub fn new(path: &str) -> Port {
+        // Create and return the port struct
+        Port {
+            socket: Arc::new(Mutex::new(PortSocket::new(path))),
```
Comment (johnnyman727) [inline tessel/src/lib.rs:88]: I think it's a little strange that the socket is thread-safe but the `ports` Box is designed to encourage developers to use it in a single-threaded manner.

File: tessel/src/lib.rs [hunk 3]
```diff
@@ -82,86 +78,146 @@ pub struct PortGroup {
...
+    pub fn pins(&mut self) -> (Pin, Pin, Pin) {
+        (
+            Pin::new(5, self.socket.clone()),
+            Pin::new(6, self.socket.clone()),
+            Pin::new(7, self.socket.clone()),
+        )
+    }
+
+    pub fn i2c<'b>(self) -> (I2cPort<'b>, Gpio<'b>) {
+        let mut available = BTreeSet::new();
```
Comment (johnnyman727) [inline tessel/src/lib.rs:101]: A `BTree` is an _interesting_ choice here! 😜

Comment (johnnyman727) [inline tessel/src/lib.rs:101]: Also, small nitpick that `available` doesn't seem like the best name. Perhaps `pins_available` would be more clear?

File: tessel/src/lib.rs [hunk 4]
```diff
@@ -82,86 +78,146 @@ pub struct PortGroup {
...
+/// Gpio is a selection of pins.
+#[allow(dead_code)]
+pub struct Gpio<'a> {
+    socket: Arc<Mutex<PortSocket>>,
+    available: BTreeSet<usize>,
+    _phantom: PhantomData<&'a Port>,
+}
 
+impl<'a> Gpio<'a> {
+    pub fn new<'b>(socket: Arc<Mutex<PortSocket>>, available: BTreeSet<usize>) -> Gpio<'b> {
```
Comment (johnnyman727) [inline tessel/src/lib.rs:118]: Rust newbie question: Why does the `new` function return a tuple with a different lifetime than the struct?

**Comment by johnnyman727** [comment]
Some comments added. 

It would also be nice to see some unit tests for `tessel`, `accel-mma84` and `climate-si2070`. Will you have time to work on those before your conf?

File: tessel/src/lib.rs (hunk 2 above)
Comment (tcr) [reply tessel/src/lib.rs:88]: Actually the `ports` function really just makes acquisition happen on one thread— it's up to the user to determine whether they want to mutex the result or not. But I'll note this in #29

File: tessel/src/lib.rs (hunk 3 above)
Comment (tcr) [reply tessel/src/lib.rs:101]: I WAS SLEEPY

File: tessel/src/lib.rs (hunk 4 above)
Comment (tcr) [reply tessel/src/lib.rs:118]: It's declaring the new method is parameterized over a new lifetime to generate Gpio (which is bounded by its own Phantomdata).

**Comment by tcr** [comment]
@johnnyman727 Probably not any unit tests JUST yet but it should be the next thing once it's worth downloading / picking up by people.

**Pull request merged**